### ISSUE TITLE
TUI to configure Meson (akin to `ccmake`)

### ISSUE THREAD

**Issue by 1ace** [mesonbuild/meson#2470]
One feature from CMake that I like is [`ccmake`](https://cmake.org/cmake/help/latest/manual/ccmake.1.html), a [TUI program](https://en.wikipedia.org/wiki/Text-based_user_interface) that allows the user to quickly see all the project's options and set the ones they want.

I would love something similar to be implemented for Meson (`meson configure --tui`, perhaps?).

(This is a feature request, and apologies if it's already been requested, I couldn't find it)

**Comment by jpakkane** [comment]
This would be nice, yes, but sadly the core developers have a lot of other stuff to do, too, so this does not have a high priority.

**Comment by 1ace** [comment]
I'll try and do it myself, but my free time is also limited :)
I started something, you can have a look on the `feature/mesonconf-tui` branch in [my fork](https://github.com/1ace/meson/tree/feature/mesonconf-tui).
I'll send a pull-request when it starts being usable :+1:

**Comment by proskur1n** [comment]
Seeing that this feature has still not been implemented, I wrote a little [tui wrapper](https://github.com/proskur1n/cmeson) around meson myself. Hopefully it will be useful for someone.

**Comment by jpakkane** [comment]
That's looking really nice.

**Comment by stephanlachnit** [comment]
I also wrote a ccmake-like TUI called [mmeson](https://github.com/stephanlachnit/mmeson). Compared to @proskur1n's excellent cmeson it has a little bit more colors but lacks some of the features (at least for now).

Both are actually quite similar as they both use [urwid](http://urwid.org/) for the TUI (and no other dependencies). I think merging something similar to Meson that could be executed via `meson configure --tui` *if* `urwid` is available (i.e. make it optional) is possible and not too difficult at this point.

The question of course is if this is even wanted upstream. I don't think a dependency-free solution is really viable since it adds a lot of unnecessary, probably unmaintained code. On the other hand Meson has the dogma of not adding non-stdlib dependencies, so any dependency has to be optional which results in the user having to use some kind of package installer anyway.

**Comment by 1ace** [comment]
I've been meaning to find the time to try writing a TUI again, this time using @willmcgugan's excellent https://github.com/textualize/textual framework, as it requires a lot less code to write and provides a lot more functionality (mainly that it's more robust). Leaving this here as a hint to anyone wanting to give it a try :slightly_smiling_face:

As for how to organize & package it, I don't know if meson already supports this but it could be done the same way as `git`, where anything installed with the name `git-foo` gets to be used as `git foo`. This would mean distros would package this new repo as something that installs `/usr/bin/meson-tui` and users would get to run it as a normal `meson tui` subcommand.

Alternatively, if the `import` of `urwid` (or `textual`) only happens in the code that gets called when the argument is passed (and has an `except` block to provide a nice message for users before re-raising the exception), distros can mark `urwid`/`textual` as an optional dependency, and those not interested can ignore it and nothing changes for them.

The downside of making it an external project (whether under `mesonbuild` or not) is obviously that it will end up having to support a wide range of meson versions, and every new meson change might break the TUI, which will be a lot of work for developers and an unreliable/bad experience for users, so you know where my vote is :wink:

**Comment by eli-schwartz** [comment]
We do have exactly one non-stdlib import (I am intentionally not counting mypy stuff that isn't imported at runtime ever).

That's tqdm, to implement a no-options "better progress reporting" when:
- generating build.ninja
- downloading wraps

The consequences of not having tqdm are none at all -- it makes no difference, affects no options or user interactions, only affects how some statuses are printed. And this is why we.have that code.

Pretty sure a --gui option doesn't qualify though.

Having a mesonbuild org project which provides it as a separate package seems fine to me, I think.

**Comment by eli-schwartz** [comment]
> The downside of making it an external project (whether under `mesonbuild` or not) is obviously that it will end up having to support a wide range of meson versions, and every new meson change might break the TUI, which will be a lot of work for developers and an unreliable/bad experience for users, so you know where my vote is 😉

I haven't taken a look at any implementations of a GUI, what do they do here? Import mesonbuild.* functions, directly access a Coredata object, etc? Use introspection json files?

**Comment by 1ace** [comment]
My implementation had split some code in mconf.py out so that it generates data structures, and then mconf.py prints those as always, and the TUI code uses those to build the UI. So 1) it requires that initial split to be merged into the meson project, and 2) any changes to what options exists (especially changes to `UserOption` and its descendents) would break the TUI if its code lives in another repo.

(One nitpick, TUI and GUI are different things: TUI is "fancy text" and can be used basically anywhere, for instance through ssh which I care about; GUI requires something like Wayland or X11, which is way heavier and less compatible, but in return allows for images and things like that)

**Comment by eli-schwartz** [comment]
I was probably somehow getting confused with cmake and forgetting that ccmake is NOT cmake-gui, despite the title being very clear. Sorry.

**Comment by dcbaker** [comment]
When I had thought about this, I wanted to use introspection data rather than altering Meson itself, given that muon (and maybe someday meson++) exist, having a TUI and/or GUI that is able to work with all of them would be great, and introspection data seems like the right way to do that, as it's also how things like meson-vscode work

**Comment by stephanlachnit** [comment]
> When I had thought about this, I wanted to use introspection data rather than altering Meson itself, given that muon (and maybe someday meson++) exist, having a TUI and/or GUI that is able to work with all of them would be great, and introspection data seems like the right way to do that, as it's also how things like meson-vscode work

I agree, this how mmeson works and iirc cmeson as well. This avoids all the version problem mentioned by @1ace, especially since they can be versioned for future changes (I already have some like min/max for int options is currently missing).
